api/layer/layer.go

```diff
@@ -470,72 +471,76 @@ func getRandomOID() (oid.ID, error) {
 	return objID, nil
 }
 
-// DeleteObject removes all objects with the passed nice name.
-func (n *layer) deleteObject(ctx context.Context, bkt *data.BucketInfo, obj *VersionedObject) *VersionedObject {
-	if len(obj.VersionID) == 0 {
-		obj.VersionID = UnversionedObjectVersionID
-	}
-	objVersion := &ObjectVersion{
-		BktInfo:               bkt,
-		ObjectName:            obj.Name,
-		VersionID:             obj.VersionID,
-		NoErrorOnDeleteMarker: true,
-	}
-
-	var nodeVersion *data.NodeVersion
-	nodeVersion, obj.Error = n.getNodeVersion(ctx, objVersion)
-
-	if obj.VersionID == UnversionedObjectVersionID {
-		if obj.Error == nil {
-			if obj.DeleteMarkVersion, obj.Error = n.removeOldVersion(ctx, bkt, nodeVersion, obj); obj.Error != nil {
-				return obj
-			}
-		} else if !errors.IsS3Error(obj.Error, errors.ErrNoSuchKey) {
+func (n *layer) deleteObject(ctx context.Context, bkt *data.BucketInfo, settings *data.BucketSettings, obj *VersionedObject) *VersionedObject {
+	if len(obj.VersionID) != 0 || settings.IsNoneStatus {
+		var nodeVersion *data.NodeVersion
+		if nodeVersion, obj.Error = n.getNodeVersionToDelete(ctx, bkt, obj); obj.Error != nil {
 			return obj
 		}
 
-		randOID, err := getRandomOID()
-		if err != nil {
-			obj.Error = fmt.Errorf("couldn't get random oid: %w", err)
+		if obj.DeleteMarkVersion, obj.Error = n.removeOldVersion(ctx, bkt, nodeVersion, obj); obj.Error != nil {
 			return obj
 		}
 
-		newVersion := &data.NodeVersion{
-			BaseNodeVersion: data.BaseNodeVersion{
-				OID:      randOID,
-				FilePath: obj.Name,
-			},
-			DeleteMarker: &data.DeleteMarkerInfo{
-				Created: time.Now(),
-				Owner:   n.Owner(ctx),
-			},
-			IsUnversioned: true,
-		}
+		obj.Error = n.treeService.RemoveVersion(ctx, bkt.CID, nodeVersion.ID)
+		return obj
+	}
 
-		if obj.Error = n.treeService.AddVersion(ctx, bkt.CID, newVersion); obj.Error != nil {
-			return obj
-		}
+	var newVersion *data.NodeVersion
 
-		n.namesCache.Delete(bkt.Name + "/" + obj.Name)
-	} else {
-		if obj.Error != nil {
-			return obj
-		}
+	if !settings.VersioningEnabled {
+		obj.VersionID = UnversionedObjectVersionID
 
-		if obj.DeleteMarkVersion, obj.Error = n.removeOldVersion(ctx, bkt, nodeVersion, obj); obj.Error != nil {
+		var nodeVersion *data.NodeVersion
+		if nodeVersion, obj.Error = n.getNodeVersionToDelete(ctx, bkt, obj); obj.Error != nil {
 			return obj
 		}
 
-		if obj.Error = n.treeService.RemoveVersion(ctx, bkt.CID, nodeVersion.ID); obj.Error != nil {
+		if obj.Error == nil {
+			if obj.DeleteMarkVersion, obj.Error = n.removeOldVersion(ctx, bkt, nodeVersion, obj); obj.Error != nil {
+				return obj
+			}
+		} else if !errors.IsS3Error(obj.Error, errors.ErrNoSuchKey) {
 			return obj
 		}
 	}
 
-	n.listsCache.CleanCacheEntriesContainingObject(obj.Name, bkt.CID)
+	randOID, err := getRandomOID()
```

Comment:
Actually, we can not reduce the umber of requests to the tree service because we have to update node anyway. The reason is to be sure that this node will be the last version (it's required to update timestamp)